### PR TITLE
Ban `ArrayToPointer` and `MutToConstPointer` from runtime MIR

### PR DESCRIPTION
**Pull request by scottmcm**
Zulip conversation: <https://rust-lang.zulipchat.com/#narrow/stream/189540-t-compiler.2Fwg-mir-opt/topic/CastKind.3A.3APointerCoercion.20in.20Runtime.20MIR/near/443955195>

Apparently MIR borrowck cares about at least one of these for checking variance.

In runtime MIR, though, there's no need for them as `PtrToPtr` does the same thing.

(Banning them simplifies passes like GVN that no longer need to handle multiple cast possibilities.)

r? mir

### CODE REVIEW

File: compiler/rustc_mir_transform/src/lib.rs [hunk 1]
```diff
@@ -506,6 +507,7 @@ fn run_analysis_cleanup_passes<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
 /// Returns the sequence of passes that lowers analysis to runtime MIR.
 fn run_runtime_lowering_passes<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
     let passes: &[&dyn MirPass<'tcx>] = &[
+        &coercions_to_casts::CoercionsToCasts,
```
Comment (cjgillot) [inline compiler/rustc_mir_transform/src/lib.rs:510]: Why not inside 'cleanup_post_borrowck' ?

**Comment by RalfJung** [comment]
> I think it's much simpler when the answer to those things is "as it says on CastKind and MirPhase::Runtime, you only need PtrToPtr here; ignore the rest".

What about grouping all of these together in a single CastKind variant `PtrToPtr(PtrCastKind)`, and declaring that op.sem can ignore the PtrCastKind? Then it's easy to consistently match on it.

This is not a string opinion though, I just feel like this PR is using the MIR validation to work around CastKind just not being of the right shape for some of its consumers.

File: compiler/rustc_middle/src/mir/syntax.rs [hunk 2]
```diff
@@ -143,6 +143,9 @@ pub enum RuntimePhase {
     /// * [`TerminatorKind::CoroutineDrop`]
     /// * [`Rvalue::Aggregate`] for any `AggregateKind` except `Array`
     /// * [`PlaceElem::OpaqueCast`]
+    /// * [`CastKind::PointerCoercion`] with any of the following:
+    ///   * [`PointerCoercion::ArrayToPointer`]
+    ///   * [`PointerCoercion::MutToConstPointer`]
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/syntax.rs:148]: compiler/rustc_const_eval/src/interpret/cast.rs should then probably also explicitly panic on these casts.

**Comment by RalfJung** [comment]
Ah, we have have this inner enum, `PointerCoercion`... but that enum is a pretty poor fit for runtime MIR in general, it is designed for coercion generation way earlier in the compiler. I wonder if we'd be better off just not using it in MIR... but I have no idea how this is relevant for borrowck.

**Comment by RalfJung** [comment]
I would expect the lifetimes in that test to be a red herring -- `mut_to_const` is the key. Someone has to reject that as a coercion (but allow it as a cast).

**Comment by RalfJung** [comment]
Ah no never mind, I shouldn't write comments when I am too tired...

on a mut-to-const implicit coercion, the type needs to stay the same, and the lifetime equality part of "the same" is checked by borrowck.

**Comment by RalfJung** [comment]
I don't have a concrete proposal either, and it seems you considered the alternatives, so my concern isn't blocking. I feel like there's got to be a better way but don't feel strongly enough about it to try and find it. ;)

**Comment by saethlin** [comment]
I'm okay with this change because it makes the runtime MIR dialect more sensible. I don't know how much this actually matters in practice, clearly it lets you clean up MIR transforms if you know what you're doing. But the fact that the enum variants still exist because all MIR dialects are the same type still makes implementing transforms confusing.

@bors r+

**Comment by fmease** [comment]
https://github.com/rust-lang/rust/pull/126669#issuecomment-2177988001
@bors r-

**Pull request merged**